### ISSUE TITLE
feature_maps export with stylegan2-ada pytorch checkpoint

### ISSUE THREAD

**Issue by losredoe132** [TJ-Solergibert/DatasetGAN-StyleGAN2#1]
Hi there, 
thank you for publishing your great work. I'm currently working on a similar project: Using the datasetgan idea with stylegan2 for the apolloscape dataset.   
I have some questions: 
1. I think you used tensorflow checkpoints from NVlabs/stylegan2 , correct? That works fine for me with official pretrained checkpoints.
2. I tried loading a stylegan2-ada pytorch checkpoint (from [here](https://github.com/NVlabs/stylegan2-ada-pytorch)). This did not work, because in line 54 of generate_feature_maps.py an exception occurs ("forward() got multiple values for argument truncation_psi")
`img = G(z, label, fm_outdir, seed, truncation_psi=truncation_psi, noise_mode=noise_mode)`
Seems like the forward() method of the pytorch implementation does not take fm_outdir argument. I looked through booth implementations but i did not understand how the feature_maps are generated from the generator of the orignal tensorflow implementation. Did you know where fm_outdir is used? I could generate the featuremaps seperatly but i would like to generate images and feature maps in parallel. 
3. Until now, inference is not implemented, right? 

Sorry for my unstructured inquiry. Thank you very much! 

Greetings 
Finn

**Comment by TJ-Solergibert** [comment]
Hello Finn!
As the title of the project states, it is still under development. But some comments about your questions:

- From 1 & 2: My initial idea was to design this project to work with StyleGAN2. NVlabs/stylegan2, NVlabs/stylegan2-ada and NVlabs/stylegan2-ada-pytorch share the same generator and (practically) discriminator architecture, so using any checkpoint trained with one of this 3 versions should work fine. Based on you comments I've noticed that when loading a NVlabs/stylegan2-ada-pytorch checkpoint produces the error you commented in 2. and it works fine with NVlabs/stylegan2. Thanks to your comment I have been able to find what is causing this error, the next few days I will work to solve it but it has to do with the way that the model is loaded. 
- From 3 + general comment: As the second line of the README states, this project is part of my final degree thesis. In the last weeks I've had problems training the model, so I have not been able to update the repo but finally yesterday I was able to generate some (good quality) synthetic data. It is planned that in the next few days I'll push a very big update including:

- [ ] A introduction and motivation of the project to the README
- [ ] A evaluation script to test the performance of the models
- [ ] A visualization script to analyze the results
- [ ] A inference script to produce synthetic labels
- [ ] Compatibility with NVlabs/stylegan2-ada-pytorch checkpoints
- [ ] Improved input pipelines with the tf.Data API

Stay tuned, feel free to comment or suggest anything both here and by the email address you will find in my profile.

Greetings
Toni

**Comment by TJ-Solergibert** [comment]
Hello Finn,

Finally yesterday I pushed a big update of the project. Re-read the README and note the following:
- It should work now with SG2-ada-pytorch checkpoints. In the first stage, check that you get correct images to confirm that.
- Some tips for labeling the images and **prototyping** your first model:
    - For creating the TFRecordDataset, you need to load a very big tensor into RAM. If you get memory errors, try removing the lower levels of the _List_of_levels_ list in the config.json file.    
    - Use LabelMe. Works fine and it's very easy to use.
    - Don't label dozens of classes. Label 5 or 6 with good precision.
    - Choose the best images your model can generate. 
    - In the _labels.txt_ file include  the _ _background_ _ and _ \__ignore__ _ class. My recommendation is to IGNORE the _ _background_ _ class (Unlabeled pixels are classified as _ _background_ _ ) and only focus on labeling the correct parts of each image. There is no necessity lo label all the pixels, maybe you are only interested in a part of the image. It's better to label 30% pixels of 30 images than 10 images to the 100%. 
    - Feel free to edit any training parameter in the _train_datasetgan.py_ script. Still needs to be improved
 
Good luck, any problem you come across, feel free to comment! For prototyping I've used a SG2 Checkpoint that generate Faces and I've been able to get pretty decent results. It's a bit tedious the part of labeling the images, but remember that quality > quantity.

**Comment by losredoe132** [comment]
Hi Toni, thank you very much for your message and support! I will try it out soon and give you feedback if it works with pytorch-ada checkpoints. Just in case it is beneficial for you to get a checkpoint from NVlabs/stylegan2 or NVlabs/stylegan2-pytorch-ada for a dataset for autonomous driving called _apolloscape_ (e.g. for out-of domain generalization experiments), feel free to contact me. Just out of interest: What data do you use?

**Comment by losredoe132** [comment]
Hey Toni, 
loading a checkpoint from /NVlabs/stylegan2-ada-pytorch works quiet well. Images are produced and label them with LabelMe is -as you said- straightforward. Converting annotation maps from .json to .npy works also without any problems. But while generating the tfrecords i run into the error shown below: 
![2022-05-24_13h09_04](https://user-images.githubusercontent.com/54853333/170021388-5fa078ce-dbe3-45e3-9646-a89829255598.png)
Casting all elemnts to the same float32 with `list_of_upsampled_fm.append(tf.cast(upsampled_fm, dytpe="float32"))` would be a solution but then i get an error in line 46 of create_TFRecords because using "/" for division is not allowed in context of tensors. So i take the .value-attribute of tensor_depth and it works fine. 

I dont think that is due to different tensorflow version, but i dont know actually.
